src/dolphin/atmosphere/ionosphere.py

```diff
@@ -0,0 +1,388 @@
+from __future__ import annotations
+
+import datetime
+import re
+from os import fspath
+from pathlib import Path
+
+import numpy as np
+import opera_utils as oput
+from rasterio.crs import CRS
+from rasterio.warp import transform_bounds
+from scipy import interpolate
+
+from dolphin import io
+from dolphin._dates import _format_date_pair, get_dates
+from dolphin._log import get_log
+from dolphin._types import Filename
+
+from .troposphere import prepare_geometry
+
+logger = get_log(__name__)
+
+###########
+
+__all__ = ["estimate_ionospheric_delay"]
+
+# constants
+K = 40.31
+SPEED_OF_LIGHT = 299792458  # meters per second
+EARTH_RADIUS = 6371.0088e3  # km
+
+
+def estimate_ionospheric_delay(
+    ifg_file_list: list[Path],
+    slc_files: dict[tuple[datetime.datetime], list[Filename]],
+    tec_files: dict[tuple[datetime.datetime], list[Filename]],
+    geom_files: list[Path],
+    output_dir: Path,
+    strides: dict[str, int] = {"x": 1, "y": 1},
+):
+    """Estimate the range delay caused by ionosphere for each interferogram.
+
+    Parameters
+    ----------
+    ifg_file_list : list[Path]
+        List of interferogram files.
+    slc_files : Dict[datetime.date, list[Filename]]
+        Dictionary of SLC files indexed by date.
+    tec_files : Dict[datetime.date, list[Filename]]
+        Dictionary of TEC files indexed by date.
+    geom_files : list[Path]
+        List of geometry files.
+    output_dir : Path
+        Output directory.
+    strides : Dict[str, int], optional
+        Strides for resampling, by default {"x": 1, "y": 1}.
+    """
+    # Read geogrid data
+    # xsize, ysize = io.get_raster_xysize(ifg_file_list[0])
+    crs = io.get_raster_crs(ifg_file_list[0])
+    epsg = crs.to_epsg()
+
+    bounds = io.get_raster_bounds(ifg_file_list[0])
+
+    if epsg != 4326:
+        lalo_bounds = transform_bounds(
+            CRS.from_epsg(epsg), CRS.from_epsg(4326), *bounds
+        )
+    else:
+        lalo_bounds = bounds
+
+    # Frame center latitude and longitude
+    latc = (lalo_bounds[1] + lalo_bounds[3]) / 2
+    lonc = (lalo_bounds[1] + lalo_bounds[3]) / 2
```

Comment:
right now these two lines are the same- since it's easy to mistype the 0/1/2/3 indexes, lets use
```suggestion
    # Frame center latitude and longitude
    latc = (top + bottom) / 2
    lonc = (left + right) / 2
```